relayer/src/message_relayer/eth_to_gear/proof_composer.rs

```diff
@@ -0,0 +1,394 @@
+};
+use ethereum_beacon_client::BeaconClient;
+use ethereum_client::{EthApi, TxHash};
+use ethereum_common::utils as eth_utils;
+use ethereum_common::{beacon, tree_hash::TreeHash, utils::MerkleProof};
+use futures::executor::block_on;
+use historical_proxy_client::traits::HistoricalProxy as _;
+use historical_proxy_client::HistoricalProxy;
+use primitive_types::H256;
+use sails_rs::{calls::Query, gclient::calls::GClientRemoting, ActorId};
+use tokio::{
+    sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender},
+    task::spawn_blocking,
+};
+use uuid::Uuid;
+
+#[derive(Clone)]
+pub struct Request {
+    pub tx: TxHashWithSlot,
+    pub tx_uuid: Uuid,
+}
+
+#[derive(Clone)]
+pub struct Response {
+    pub payload: EthToVaraEvent,
+    pub tx_uuid: Uuid,
+}
+
+pub struct ProofComposerIo {
+    requests_channel: UnboundedSender<Request>,
+    responses_channel: UnboundedReceiver<Response>,
+}
+
+impl ProofComposerIo {
+    pub fn new(
+        requests_channel: UnboundedSender<Request>,
+        responses_channel: UnboundedReceiver<Response>,
+    ) -> Self {
+        Self {
+            requests_channel,
+            responses_channel,
+        }
+    }
+
+    /// Receive composed proof for some transaction.
+    ///
+    /// In case of `None` indicates closed channel.
+    pub async fn recv(&mut self) -> Option<Response> {
+        self.responses_channel.recv().await
+    }
+
+    /// Send request to compose proof for `tx` with uuid `tx_uuid`.
+    ///
+    /// Returns `false` if send failed which indicates that channel was closed.
+    pub fn compose_proof_for(&mut self, tx_uuid: Uuid, tx: TxHashWithSlot) -> bool {
+        self.requests_channel.send(Request { tx, tx_uuid }).is_ok()
+    }
+}
+
+pub struct ProofComposer {
+    pub api_provider: ApiProviderConnection,
+    pub beacon_client: BeaconClient,
+    pub eth_api: EthApi,
+    pub waiting_for_checkpoints: Vec<(Uuid, TxHashWithSlot)>,
+    pub last_checkpoint: Option<EthereumSlotNumber>,
+    pub historical_proxy_address: H256,
+    pub suri: String,
+}
+
+impl ProofComposer {
+    pub fn new(
+        api_provider: ApiProviderConnection,
+        beacon_client: BeaconClient,
+        eth_api: EthApi,
+        historical_proxy_address: H256,
+        suri: String,
+    ) -> Self {
+        Self {
+            api_provider,
+            beacon_client,
+            eth_api,
+            waiting_for_checkpoints: Vec::new(),
+            last_checkpoint: None,
+            historical_proxy_address,
+            suri,
+        }
+    }
+
+    pub fn run(self, checkpoints: UnboundedReceiver<EthereumSlotNumber>) -> ProofComposerIo {
+        let (requests_tx, requests_rx) = unbounded_channel();
+        let (response_tx, response_rx) = unbounded_channel();
+
+        spawn_blocking(move || {
+            block_on(task(self, checkpoints, requests_rx, response_tx));
+        });
+
+        ProofComposerIo::new(requests_tx, response_rx)
+    }
+
+    async fn run_inner(
```

Comment:
perhaps it is worth to make this method into standalone function similar to the `task` method to decrease nesting